File: LNCollectionView/LNScrollView/Simulator/LNScrollViewDecelerateSimulator.m
//
//  LNScrollViewDecelerateSimulator.m
//  LNCollectionView
//
//  Created by Levison on 7.11.24.
//

#import "LNScrollViewDecelerateSimulator.h"

@interface LNScrollViewDecelerateSimulator ()

@property (nonatomic, assign) CGFloat position;
@property (nonatomic, assign) CGFloat velocity;

@property (nonatomic, assign) CGFloat damping;
@property (nonatomic, assign) NSTimeInterval currentTime;

@end

@implementation LNScrollViewDecelerateSimulator

- (instancetype)initWithPosition:(CGFloat)position velocity:(CGFloat)velocity
{
    self = [super init];
    if (self) {
        self.damping = 2.f;
        self.position = position;
        self.velocity = velocity;
    }
    return self;
}

- (void)accumulate:(NSTimeInterval)during
{
    self.currentTime += during;
    CGFloat v = self.velocity * exp(- self.damping * during);
    CGFloat l = (-1.f/self.damping) * self.velocity * exp(- self.damping * during) - (-1.f/self.damping) * self.velocity;
    if (self.velocity < 0.01) {
        self.velocity = 0;
    }
    self.velocity = v;
    self.position = self.position + l;
}

- (BOOL)isFinished
{
    return fabs(self.velocity) < 0.01;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/scrollview/simulator/LNScrollViewDecelerateSimulator.kt
package com.ln.collectionview.scrollview.simulator

class LNScrollViewDecelerateSimulator(
    initialPosition: Float,
    initialVelocity: Float
) {

    var position: Float = initialPosition
        private set
    var velocity: Float = initialVelocity
        private set

    private var damping: Float = 2f
    private var currentTime: Long = 0 // 当前时间，以毫秒为单位

    fun accumulate(during: Long) {
        currentTime += during
        val v = velocity * Math.exp((-damping * during / 1000.0).toDouble()).toFloat()
        val l = (-1f / damping) * velocity * Math.exp((-damping * during / 1000.0).toDouble()).toFloat() - (-1f / damping) * velocity
        if (Math.abs(velocity) < 0.01f) {
            velocity = 0f
        }
        velocity = v
        position += l
    }

    fun isFinished(): Boolean {
        return Math.abs(velocity) < 0.01f
    }
}
